File: backend/app/Models/DietEntry.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class DietEntry extends Model
{
    use HasFactory;

    protected $fillable = [
        'user_id', 'food_id', 'quantity', 'date', 'meal_type'
    ];

    protected $casts = [
        'date'     => 'date',
        'quantity' => 'float',
    ];

    public function food(): BelongsTo
    {
        return $this->belongsTo(Food::class);
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    public function calculateNutrients(): array
    {
        $food  = $this->food;
        $ratio = $this->quantity / 100;

        return [
            'id'            => $this->id,
            'food_id'       => $this->food_id,
            'food_name'     => $food->name,
            'quantity'      => $this->quantity,
            'date'          => $this->date->format('Y-m-d'),
            'meal_type'     => $this->meal_type,
            'calories'      => round($food->calories * $ratio),
            'protein'       => round($food->protein * $ratio, 1),
            'fat'           => round($food->fat * $ratio, 1),
            'carbohydrates' => round($food->carbohydrates * $ratio, 1)
        ];
    }
}

File: mobile/lib/data/models/diet/diet_entry.dart
class DietEntry {
  final int? id;
  final int foodId;
  final String foodName;
  final double quantity;
  final String date;
  final String mealType;
  final int calories;
  final double protein;
  final double fat;
  final double carbohydrates;

  DietEntry({
    this.id,
    required this.foodId,
    required this.foodName,
    required this.quantity,
    required this.date,
    required this.mealType,
    required this.calories,
    required this.protein,
    required this.fat,
    required this.carbohydrates,
  });

  factory DietEntry.fromJson(Map<String, dynamic> json) {
    return DietEntry(
      id: json['id'],
      foodId: json['food_id'],
      foodName: json['food_name'] ?? '',
      quantity: json['quantity'].toDouble(),
      date: json['date'],
      mealType: json['meal_type'],
      calories: json['calories'],
      protein: json['protein'].toDouble(),
      fat: json['fat'].toDouble(),
      carbohydrates: json['carbohydrates'].toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {
      'food_id': foodId,
      'quantity': quantity,
      'date': date,
      'meal_type': mealType,
    };

    if (id != null) {
      data['id'] = id;
    }

    return data;
  }

  DietEntry copyWith({
    int? id,
    int? foodId,
    String? foodName,
    double? quantity,
    String? date,
    String? mealType,
    int? calories,
    double? protein,
    double? fat,
    double? carbohydrates,
  }) {
    return DietEntry(
      id: id ?? this.id,
      foodId: foodId ?? this.foodId,
      foodName: foodName ?? this.foodName,
      quantity: quantity ?? this.quantity,
      date: date ?? this.date,
      mealType: mealType ?? this.mealType,
      calories: calories ?? this.calories,
      protein: protein ?? this.protein,
      fat: fat ?? this.fat,
      carbohydrates: carbohydrates ?? this.carbohydrates,
    );
  }
}